src/Features/CSharp/Portable/CodeFixes/AddExplicitCast/AddExplicitCastCodeFixProvider.cs

```diff
@@ -0,0 +1,398 @@
+                        context.Diagnostics);
+                }
+                else if (hasSolution && potentialConversionTypes.Length > 1)
+                {
+                    var actions = ArrayBuilder<CodeAction>.GetInstance();
+
+                    // MaximumConversionOptions: we show at most [MaximumConversionOptions] options for this code fixer
+                    for (var i = 0; i < Math.Min(MaximumConversionOptions, potentialConversionTypes.Length); i++)
+                    {
+                        var convType = potentialConversionTypes[i];
+                        actions.Add(new MyCodeAction(string.Format(CSharpFeaturesResources.Convert_type_to_0, convType.ToMinimalDisplayString(semanticModel, context.Span.Start)),
+                            c => ApplyFixAsync(context.Document, root, targetNode, convType)));
+                    }
+
+                    if (potentialConversionTypes.Length > MaximumConversionOptions)
+                    {
+                        // If the number of potential conversion types is larger than options we could show, report telemetry
+                        Logger.Log(FunctionId.CodeFixes_AddExplicitCast,
+                            KeyValueLogMessage.Create(m =>
+                            {
+                                m["NumberOfCandidates"] = potentialConversionTypes.Length;
+                            }));
+                    }
+
+                    context.RegisterCodeFix(new CodeAction.CodeActionWithNestedActions(
+                        CSharpFeaturesResources.Add_explicit_cast,
+                        actions.ToImmutableAndFree(), isInlinable: false),
+                        context.Diagnostics);
+                }
+            }
+        }
+
+        private static Task<Document> ApplyFixAsync(Document document, SyntaxNode currentRoot, ExpressionSyntax targetNode, ITypeSymbol conversionType)
+        {
+            var castExpression = targetNode.Cast(conversionType);
+            var newRoot = currentRoot.ReplaceNode(targetNode, castExpression.WithAdditionalAnnotations(Simplifier.Annotation));
+            return Task.FromResult(document.WithSyntaxRoot(newRoot));
+        }
+
+        /// <summary>
+        /// Output the current type information of the target node and the conversion type(s) that the target node is going to be cast by.
+        /// Implicit downcast can appear on Variable Declaration, Return Statement, and Function Invocation
+        /// </summary>
+        /// For example:
+        /// Base b; Derived d = [||]b;       
+        /// "b" is the current node with type "Base", and the potential conversion types list which "b" can be cast by is {Derived}
+        /// 
+        /// root: The root of the tree of nodes.
+        /// targetNode: The node to be cast.
+        /// targetNodeType: Output the type of "targetNode".
+        /// potentialConversionTypes: Output the potential conversions types that "targetNode" can be cast to
```

Comment:
Those can be doc using `<param name="...">...</param>`

Reply:
Didn't see Andrew's comments above since I disabled comments in codeflow. You can ignore this. It was just meant to be an FYI

---
In reply to: [387349461](https://github.com/dotnet/roslyn/pull/41371#discussion_r387349461) [](ancestors = 387349461)